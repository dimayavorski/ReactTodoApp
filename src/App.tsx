import React from 'react';
import logo from './logo.svg';
import { AiOutlinePlus } from 'react-icons/ai'
import './App.css';

function App() {
  return (
    <div className="App">
      <h3>TODO App</h3>
      <form>
        <input type="text" className="" placeholder="Add todo"/>
        <button><AiOutlinePlus/></button>
      </form>
    </div>
  );
}

export default App;
