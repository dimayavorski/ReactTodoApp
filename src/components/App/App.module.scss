.app {
	background-color: aqua;
	padding: 15px 15px 15px 15px;

	.title {
		font-size: 20px;
	}

	.inputContainer {
		display: flex;
		justify-content: space-between;
		padding: 15px;
		border-radius: 8px;
		background-color: white;

		.input {
			width: 100%;

			border: none;
			font-size: 20px;
			line-height: 20px;

			&:focus-visible {
				outline: none;
			}
		}

		.addTodo {
			background: plum;
			color: white;
			border-radius: 15px;
			border: none;
			border-color: plum;
		}
	}
}

@media screen and (min-width: 768px) {
	.app {
		padding: 15px 50px 15px 50px;

		.title {
			font-size: 25px;
		}
	}
}

@media screen and (min-width: 992px) {
	.app {
		padding: 15px 100px 15px 100px;

		.title {
			font-size: 30px;
		}
	}
}

@media screen and (min-width: 1200px) {
	.app {
		padding: 15px 350px 15px 350px;

		.title {
			font-size: 40px;
		}
	}
}

@media screen and (min-width: 1600px) {
	.app {
		padding: 15px 500px 15px 500px;
	}
}
