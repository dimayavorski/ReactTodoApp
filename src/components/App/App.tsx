import { AiOutlinePlus } from 'react-icons/ai';
import styles from './App.module.scss';
import { ToDoList } from '../ToDoList/ToDoList';
import { useState } from 'react';
import { useTodos } from '../../hooks/useTodos';
import { Checkbox } from '../Checkbox/Checkbox';
import { Header } from '../Header/Header';

function App() {
	const [text, setText] = useState('');

	const [isActive, setIsActive] = useState<boolean>(false);
	const { add } = useTodos();

	const toggleCheckbox = (value: boolean) => {
		setIsActive(value);
	};
	const submitHandler = (event: React.FormEvent) => {
		event.preventDefault();
		if (text.trim().length > 0) {
			setText('');
			setIsActive(false);
			add(text, isActive);
		}
	};

	return (
		<div className={styles.app}>
			<Header />

			<form onSubmit={submitHandler}>
				<div className={styles.inputContainer} data-testid="inputContainer">
					<Checkbox
						isActive={isActive}
						toggleCheckboxHandler={toggleCheckbox}
					/>
					<input
						type="text"
						className={styles.input}
						placeholder="Add todo"
						value={text}
						onChange={(e) => setText(e.target.value)}
					/>
					<button type="submit" disabled={!text} className={styles.addTodo}>
						<AiOutlinePlus size={30} />
					</button>
				</div>
			</form>
			<ToDoList />
		</div>
	);
}

export default App;
