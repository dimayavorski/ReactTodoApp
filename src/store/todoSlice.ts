import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { ITodoItem } from '../models/ITodoItem';
import { FilterType } from '../models/FilterType';

const todos: ITodoItem[] = [];
const filter: number = FilterType.ALL;

const todoSlice = createSlice({
	name: 'todos',
	initialState: {
		todos,
		filter,
	},
	reducers: {
		addToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos.push(action.payload);
		},
		removeToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos = state.todos.filter((todo) => todo.id !== action.payload.id);
		},
		toggleToDoComplete(state, action: PayloadAction<ITodoItem>) {
			const toggleTodo = state.todos.find(
				(todo) => todo.id === action.payload.id
			);
			if (toggleTodo) toggleTodo.checked = !toggleTodo.checked;
		},
		updateToDos(state, action: PayloadAction<ITodoItem[]>) {
			state.todos = action.payload;
		},
		updateFilter(state, action: PayloadAction<FilterType>) {
			state.filter = action.payload;
		},
	},
});

export const {
	updateToDos,
	addToDo,
	removeToDo,
	toggleToDoComplete,
	updateFilter,
} = todoSlice.actions;

export default todoSlice.reducer;
