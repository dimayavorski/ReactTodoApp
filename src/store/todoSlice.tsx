import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { ITodoItem } from '../models/ITodoItem';

const todos: ITodoItem[] = [];

const todoSlice = createSlice({
	name: 'todos',
	initialState: {
		todos,
	},
	reducers: {
		addToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos.push(action.payload);
		},
		removeToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos = state.todos.filter((todo) => todo.id !== action.payload.id);
		},
		toggleToDoComplete(state, action: PayloadAction<ITodoItem>) {
			const toggleTodo = state.todos.find(
				(todo) => todo.id === action.payload.id
			);
			if (toggleTodo) toggleTodo.checked = !toggleTodo.checked;
		},
		updateToDos(state, action: PayloadAction<ITodoItem[]>) {
			state.todos = action.payload;
		},
	},
});

export const { updateToDos, addToDo, removeToDo, toggleToDoComplete } =
	todoSlice.actions;
//export const selectCount = (state: RootState) => state.counter.value
export default todoSlice.reducer;
