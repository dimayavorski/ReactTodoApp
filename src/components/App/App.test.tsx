import {
	RenderResult,
	fireEvent,
	render,
	screen,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import App from './App';
import { store } from '../../store/store';

describe('App tests', () => {
	let component: RenderResult;

	beforeEach(() => {
		component = render(
			<Provider store={store}>
				<App />
			</Provider>
		);
	});

	it('should match snapshot', () => {
		expect(component).toMatchSnapshot();
	});
	it('should render component', () => {
		expect(screen.getByPlaceholderText('Add todo')).toBeInTheDocument();
	});

	it('should create new todo', () => {
		const inputElement = screen.getByPlaceholderText('Add todo');
		const saveBtn = component.container.getElementsByClassName('addTodo');

		expect(saveBtn.length).toBe(1);
		fireEvent.change(inputElement, { target: { value: 'Test Title' } });
		fireEvent.click(saveBtn[0]);

		expect(screen.getByText('Test Title')).toBeInTheDocument();
	});

	it('should create todo with empty title', () => {
		const inputElement = screen.getByPlaceholderText('Add todo');
		const saveBtn = component.container.getElementsByClassName('addTodo');

		expect(saveBtn.length).toBe(1);
		fireEvent.change(inputElement, { target: { value: '       ' } });
		fireEvent.click(saveBtn[0]);

		expect(screen.queryByText('       ')).not.toBeInTheDocument();
	});
});
