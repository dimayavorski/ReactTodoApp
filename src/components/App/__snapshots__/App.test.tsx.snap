// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app"
      >
        <h3
          class="title"
        >
          TODO App
        </h3>
        <form>
          <div
            class="inputContainer"
          >
            <div
              class="checkBoxContainer"
            >
              <label>
                <input
                  type="checkbox"
                />
                <span />
              </label>
            </div>
            <input
              class="input"
              placeholder="Add todo"
              type="text"
              value=""
            />
            <button
              class="addTodo"
            >
              <svg
                fill="currentColor"
                height="30"
                stroke="currentColor"
                stroke-width="0"
                t="1551322312294"
                version="1.1"
                viewBox="0 0 1024 1024"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <defs />
                <path
                  d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                />
                <path
                  d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                />
              </svg>
            </button>
          </div>
        </form>
        <div
          class="toDoList"
        >
          <div
            class="content"
          >
            <ul
              style="overflow-y: auto; height: 100%; list-style-type: none; padding-left: 0px; cursor: pointer;"
            />
          </div>
          <ul
            class="actionsList"
          >
            <li>
              0
               items left
            </li>
            <li
              class="filter"
            >
              <span>
                All
              </span>
              <span>
                Active
              </span>
              <span>
                Completed
              </span>
            </li>
            <li>
              <span>
                Clear Completed
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app"
    >
      <h3
        class="title"
      >
        TODO App
      </h3>
      <form>
        <div
          class="inputContainer"
        >
          <div
            class="checkBoxContainer"
          >
            <label>
              <input
                type="checkbox"
              />
              <span />
            </label>
          </div>
          <input
            class="input"
            placeholder="Add todo"
            type="text"
            value=""
          />
          <button
            class="addTodo"
          >
            <svg
              fill="currentColor"
              height="30"
              stroke="currentColor"
              stroke-width="0"
              t="1551322312294"
              version="1.1"
              viewBox="0 0 1024 1024"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs />
              <path
                d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
              />
              <path
                d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
              />
            </svg>
          </button>
        </div>
      </form>
      <div
        class="toDoList"
      >
        <div
          class="content"
        >
          <ul
            style="overflow-y: auto; height: 100%; list-style-type: none; padding-left: 0px; cursor: pointer;"
          />
        </div>
        <ul
          class="actionsList"
        >
          <li>
            0
             items left
          </li>
          <li
            class="filter"
          >
            <span>
              All
            </span>
            <span>
              Active
            </span>
            <span>
              Completed
            </span>
          </li>
          <li>
            <span>
              Clear Completed
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
